### PR TITLE
CDAP-13430 run sinks in parallel for spark pipelines

### PR DESCRIPTION
**Pull request by albertshau**

### CODE REVIEW

**Comment by albertshau** [comment]
https://builds.cask.co/browse/CDAP-RUT1503

File: cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/SparkCollection.java [hunk 1]
```diff
@@ -58,9 +58,9 @@
 
   <U> SparkCollection<U> compute(StageSpec stageSpec, SparkCompute<T, U> compute) throws Exception;
 
-  void store(StageSpec stageSpec, PairFlatMapFunction<T, Object, Object> sinkFunction);
+  Runnable store(StageSpec stageSpec, PairFlatMapFunction<T, Object, Object> sinkFunction);
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/SparkCollection.java]: Seems like this method is no longer doing the `store`. So maybe better renaming it to `createStoreTask`.

File: cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/SparkPipelineRunner.java [hunk 2]
```diff
@@ -329,6 +336,18 @@ public void runPipeline(PipelinePhase pipelinePhase, String sourcePluginType,
...
+
+    Collection<Future> sinkFutures = new ArrayList<>(sinkRunnables.size());
+    ExecutorService executorService = Executors.newFixedThreadPool(sinkRunnables.size(), new ThreadFactoryBuilder()
+      .setNameFormat("pipeline-sink-task")
+      .build());
+    for (Runnable runnable : sinkRunnables) {
+      sinkFutures.add(executorService.submit(runnable));
+    }
+    for (Future future : sinkFutures) {
+      future.get();
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/SparkPipelineRunner.java]: If the task of this future failed, the `get()` call would throw, right? Meaning other tasks might still be running and there is no cancellation of them, and neither the executor service is being shutdown. I think it's better to have a `try-catch` around the `future.get()` and accumulate the exceptions and throw it out after all tasks are completed. Or if there is way to cancel the unfinished tasks, we should cancel them before throwing exception.

File: cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/SparkPipelineRunner.java [hunk 3]
```diff
@@ -329,6 +342,42 @@ public void runPipeline(PipelinePhase pipelinePhase, String sourcePluginType,
...
+      sinkFutures.add(executorService.submit(runnable));
+    }
+
+    Throwable error = null;
+    Iterator<Future> futureIter = sinkFutures.iterator();
+    while (futureIter.hasNext()) {
+      Future future = futureIter.next();
+      try {
+        future.get();
+      } catch (ExecutionException | InterruptedException e) {
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/SparkPipelineRunner.java]: If the exception is `InterruptedException`, even the task of the current `future` is not yet completed. The interruption is on the `get` call, not on the task itself.

File: cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/SparkPipelineRunner.java [hunk 4]
```diff
@@ -329,6 +342,42 @@ public void runPipeline(PipelinePhase pipelinePhase, String sourcePluginType,
...
+        }
+        if (error == null) {
+          error = cause;
+        } else {
+          error.addSuppressed(cause);
+        }
+        LOG.error("A spark job failed, cancelling all other jobs.");
+        while (futureIter.hasNext()) {
+          Future otherFuture = futureIter.next();
+          otherFuture.cancel(true);
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/SparkPipelineRunner.java]: If you call `shutdownNow` on the executor service, the effect is the same as this loop. Probably don't need this loop.

**Comment by chtyim** [comment]
LGTM

**Pull request merged**